src/limiter/bound_preserving_limiter.cpp

```diff
@@ -0,0 +1,1003 @@
+                }
+
+                if (nstate == dim + 2) {
+                    std::vector<real> t2(n_quad_pts, 1);
+                    real theta2 = 1.0;
+                    std::array<real, nstate> soln_at_iquad = {};
+
+                    // Obtain theta2 value
+                    for (unsigned int iquad = 0; iquad < n_quad_pts; ++iquad) {
+                        for (unsigned int istate = 0; istate < nstate; ++istate) {
+                            soln_at_iquad[istate] = soln_at_q[istate][iquad];
+                        }
+                        real p_lim = 0;
+
+                        if (nstate == dim + 2)
+                            p_lim = euler_physics->compute_pressure(soln_at_iquad);
+
+                        if (p_lim >= 0)
+                            t2[iquad] = 1;
+                        else
+                            t2[iquad] = p_avg / (p_avg - p_lim);
+
+                        if (t2[iquad] != 1) {
+                            if (t2[iquad] >= 0 && t2[iquad] <= 1 && t2[iquad] < theta2) {
+                                theta2 = t2[iquad];
+                            }
+                        }
+                    }
+
+                    // Limit values at quadrature points
+                    for (unsigned int istate = 0; istate < nstate; ++istate) {
+                        for (unsigned int iquad = 0; iquad < n_quad_pts; ++iquad) {
+                            soln_at_q[istate][iquad] = theta2 * (soln_at_q[istate][iquad] - soln_cell_avg[istate])
+                                + soln_cell_avg[istate];
+                        }
+                    }
+                }
+
+                // Project soln at quadrature points to dofs.
+                for (int istate = 0; istate < nstate; istate++) {
+                    soln_basis_projection_oper.matrix_vector_mult_1D(soln_at_q[istate], soln_dofs[istate], soln_basis_projection_oper.oneD_vol_operator);
+                }
+
+                // Write limited solution dofs to the global solution vector.
+                for (int istate = 0; istate < nstate; istate++) {
+                    for (unsigned int ishape = 0; ishape < n_shape_fns; ++ishape) {
+                        const unsigned int idof = istate * n_shape_fns + ishape;
+                        solution[current_dofs_indices[idof]] = soln_dofs[istate][ishape]; //
+
+                        // Verify density is still positive after theta2 limiter is applied
+                        if (istate == 0) {
+                            if (solution[current_dofs_indices[idof]] < 0) {
+                                std::cout << "Density is a negative value - Aborting... " << std::endl << solution[current_dofs_indices[idof]] << std::endl << std::flush;
+                                std::abort();
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        template class BoundPreservingLimiter <PHILIP_DIM, double>;
```

Comment:
I've made separate cpp/headers files for all the limiters, will push the changes soon, but I just wanted to ask if it would it be alright to keep the two different version of the positivity preserving limiter in a single cpp/header file since they are very similar?

Reply:
Sure, if they are very similar, it usually means that they have similar codes that can be reused. In that scenario, you might have some common supporting functions that can be defined in the .cpp's [anonymous](https://stackoverflow.com/questions/6598244/private-static-member-function-or-free-function-in-anonymous-namespace) namespace.

Reply:
Separated all the classes into their own cpp/header files. Since the positivity limiters are extremely similar, I ended up making the positivity limiter into one class.